rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -52,10 +55,92 @@ NodeParameters::NodeParameters(
     use_intra_process,
     allocator);
 
+  // Get the node options
+  const rcl_node_t * node = node_base->get_rcl_node_handle();
+  if (NULL == node) {
+    throw std::runtime_error("Need valid node handle in NodeParameters");
+  }
+  const rcl_node_options_t * options = rcl_node_get_options(node);
+  if (NULL == options) {
+    throw std::runtime_error("Need valid node options NodeParameters");
+  }
+
+  // Get paths to yaml files containing initial parameter values
+  std::vector<std::string> yaml_paths;
+
+  auto get_yaml_paths = [&yaml_paths, &options](const rcl_arguments_t * args) {
+      int num_yaml_files = rcl_arguments_get_param_files_count(args);
+      if (num_yaml_files > 0) {
+        char ** param_files;
+        rcl_ret_t ret = rcl_arguments_get_param_files(args, options->allocator, &param_files);
+        if (RCL_RET_OK != ret) {
+          rclcpp::exceptions::throw_from_rcl_error(ret);
+        }
+        for (int i = 0; i < num_yaml_files; ++i) {
+          yaml_paths.emplace_back(param_files[i]);
+        }
+      }
+    };
+
+  // global before local so that local overwrites global
+  if (options->use_global_arguments) {
+    get_yaml_paths(rcl_get_global_arguments());
+  }
+  get_yaml_paths(&(options->arguments));
+
+  // Get fully qualified node name post-remapping to use to find node's params in yaml files
+  const std::string node_name = node_base->get_name();
+  const std::string node_namespace = node_base->get_namespace();
+  if (0u == node_namespace.size() || 0u == node_name.size()) {
+    // Should never happen
+    throw std::runtime_error("Node name and namespace were not set");
+  }
+  std::string combined_name;
+  if ('/' == node_namespace.at(node_namespace.size() - 1)) {
+    combined_name = node_namespace + node_name;
+  } else {
+    combined_name = node_namespace + '/' + node_name;
+  }
+
+  std::map<std::string, rclcpp::Parameter> parameters;
+
+  // TODO(sloretz) rcl too parse yaml when circular dependency is solved
+  for (const std::string & yaml_path : yaml_paths) {
+    rcl_params_t * yaml_params = rcl_yaml_node_struct_init(options->allocator);
+    if (NULL == yaml_params) {
+      throw std::runtime_error("Failed to initialize yaml params struct");
```

Comment:
Should be able to narrow this down to a better exception, like `std::bad_alloc`. Looking at them now, the yaml parsing API in yaml leaves something to be desired in terms of error reporting and namespacing.

Reply:
Changed this exception to `bad_alloc` in ef427a7